File: leetcode/8_binary_search/1760_Minimum_Limit_of_Balls_in_a_Bag.py
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        def _can_divide(max_balls):
            ops = 0
            for n in nums:
                ops += ceil(n / max_balls) - 1
                if ops > maxOperations:
                    return False
            return True

        left, right = 1, max(nums)
        while left < right:
            mid = left + ((right - left) // 2)
            if _can_divide(mid):
                right = mid
            else:
                left = mid + 1
        return left

File: leetcode/8_binary_search/1760_Minimum_Limit_of_Balls_in_a_Bag.go
func canDivide(nums []int, maxBalls int, maxOperations int) bool {
	ops := 0
	for _, n := range nums {
		ops += (n+maxBalls-1)/maxBalls - 1
		if ops > maxOperations {
			return false
		}
	}
	return true
}

func minimumSize(nums []int, maxOperations int) int {
	left, right := 1, 0
	for _, n := range nums {
		if n > right {
			right = n
		}
	}
	res := right

	for left < right {
		mid := left + (right-left)/2
		if canDivide(nums, mid, maxOperations) {
			right = mid
			res = right
		} else {
			left = mid + 1
		}
	}
	return res
}
